File: backends/dotnet/src/Zappy.Adapters.GraphQL/Types/OrderConnection.cs
using Zappy.Application;
using Zappy.Domain;

namespace Zappy.Adapters.GraphQL;

public sealed class OrderConnection(IReadOnlyList<OrderEdge> edges, PageInfo pageInfo, int totalCount)
{
    public IReadOnlyList<OrderEdge> Edges { get; } = edges;

    public PageInfo PageInfo { get; } = pageInfo;

    public int TotalCount { get; } = totalCount;

    public static OrderConnection From(Page<Order> page)
    {
        var edges = page.Items.Select(order => new OrderEdge(Cursor.For(order.Id), order)).ToList();
        return new OrderConnection(
            edges,
            new PageInfo(page.HasNextPage, edges.Count == 0 ? null : edges[^1].Cursor),
            page.TotalCount);
    }
}

File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/graphql/OrderConnection.java
package com.zappymart.adapters.graphql;

import com.zappymart.application.Page;
import com.zappymart.domain.ordering.Order;

import java.util.List;

public record OrderConnection(List<OrderEdge> edges, PageInfo pageInfo, int totalCount) {

    public static OrderConnection of(Page<Order> page) {
        List<OrderEdge> edges = page.items().stream()
                .map(order -> new OrderEdge(Cursors.encode(order.id()), order))
                .toList();
        String endCursor = edges.isEmpty() ? null : edges.getLast().cursor();
        return new OrderConnection(edges, new PageInfo(page.hasNextPage(), endCursor), page.totalCount());
    }
}
